common/persistence/sql/sqlMatchingPersistence.go

```diff
@@ -772,118 +950,282 @@ func createCurrentExecution(tx *sqlx.Tx, request *persistence.CreateWorkflowExec
-		return &workflow.InternalServiceError{
-			Message: fmt.Sprintf("Failed to lock executions row to check next_event_id. Error: %v", err),
+func createTimerTasks(tx *sqlx.Tx, timerTasks []persistence.Task, deleteTimerTask persistence.Task, shardID int, domainID, workflowID, runID string) error {
+	if len(timerTasks) > 0 {
+		structs := make([]timerTasksRow, len(timerTasks))
+
+		for i, task := range timerTasks {
+			switch t := task.(type) {
+			case *persistence.DecisionTimeoutTask:
+				structs[i].EventID = t.EventID
+				structs[i].TimeoutType = t.TimeoutType
+				structs[i].ScheduleAttempt = t.ScheduleAttempt
+			case *persistence.ActivityTimeoutTask:
+				structs[i].EventID = t.EventID
+				structs[i].TimeoutType = t.TimeoutType
+				structs[i].ScheduleAttempt = t.Attempt
+			case *persistence.UserTimerTask:
+				structs[i].EventID = t.EventID
+			case *persistence.RetryTimerTask:
+				structs[i].EventID = t.EventID
+				structs[i].ScheduleAttempt = int64(t.Attempt)
+			}
+
+
+			structs[i].ShardID = shardID
+			structs[i].DomainID = domainID
+			structs[i].WorkflowID = workflowID
+			structs[i].RunID = runID
+
+			structs[i].VisibilityTimestamp = persistence.GetVisibilityTSFrom(task)
+			structs[i].TaskID = task.GetTaskID()
+			structs[i].Version = task.GetVersion()
+			structs[i].TaskType = task.GetType()
+		}
+
+		query, args, err := tx.BindNamed(createTimerTasksSQLQuery, structs)
+		if err != nil {
+			return &workflow.InternalServiceError{
+				Message: fmt.Sprintf("Failed to create timer tasks. Failed to bind query. Error: %v", err),
+			}
+		}
+
+		if result, err := tx.Exec(query, args...); err != nil {
+			return &workflow.InternalServiceError{
+				Message: fmt.Sprintf("Failed to create timer tasks. Error: %v", err),
+			}
+		} else {
+			rowsAffected, err := result.RowsAffected()
+			if err != nil {
+				return &workflow.InternalServiceError{
+					Message: fmt.Sprintf("Failed to create timer tasks. Could not verify number of rows inserted. Error: %v", err),
+				}
+			}
+
+			if int(rowsAffected) != len(timerTasks) {
+				return &workflow.InternalServiceError{
+					Message: fmt.Sprintf("Failed to create timer tasks. Inserted %v instead of %v rows into timer_tasks. Error: %v", rowsAffected, len(timerTasks), err),
+				}
+			}
 		}
 	}
-	if nextEventID != condition {
-		return &persistence.ConditionFailedError{
-			Msg: fmt.Sprintf("next_event_id was %v when it should have been %v.", nextEventID, condition),
+
+	if deleteTimerTask != nil {
+		if _, err := tx.Exec(completeTimerTaskSQLQuery, shardID, persistence.GetVisibilityTSFrom(deleteTimerTask), deleteTimerTask.GetTaskID());
+			err != nil {
+			return &workflow.InternalServiceError{
+				Message: fmt.Sprintf("Failed to delete timer task. Task: %v. Error: %v", deleteTimerTask, err),
+			}
 		}
 	}
+
 	return nil
 }
+
+func prepareLastReplicationInfoForSerialization(m map[string]*persistence.ReplicationInfo) map[string]persistence.ReplicationInfo {
```

Comment:
I don't think this is necessary. I created the following test and it's working.
```
func TestGobSerialization(t *testing.T) {
	value := make(map[string]*persistence.ReplicationInfo)

	value["test"] = &persistence.ReplicationInfo{Version:1, LastEventID:2}

	bytes, err := gobSerialize(value)
	require.NoError(t, err)

	var result map[string]*persistence.ReplicationInfo
	err = gobDeserialize(bytes, &result)
	require.NoError(t, err)
}
```

Reply:
i'm concerned about this because if i serialize a pointer to an object, the object will probably not exist at that address when deserializing at a later time. i'm not sure whether gobbing a pointer gobs the dereferenced object

in your test both s11n/des11n are in the same execution of a single program so probably there is no risk of the ReplicationInfo being gc'd, much less not even existing

Reply:
indeed you are right, pointers get dereferenced first. here is an example (the byte array came from a `map[int]*A`)

```
package main

import (
	"encoding/gob"
	"bytes"
	"fmt"
)

type A struct {
 X int
 Y int
}

func main() {

var b = bytes.NewBuffer([]byte{15,255,131,4,1,2,255,132,0,1,4,1,255,130,0,0,24,255,129,3,1,2,255,130,0,1,2,1,1,88,1,4,0,1,1,89,1,4,0,0,0,19,255,132,0,2,4,1,44,1,254,1,188,0,2,1,22,1,255,222,0})

var a map[int]*A

dec := gob.NewDecoder(b)
dec.Decode(&a)

fmt.Println(a)
fmt.Println(a[1])

}
```